datafusion/physical-expr/src/scalar_function.rs

```diff
@@ -177,7 +177,18 @@ impl PhysicalExpr for ScalarFunctionExpr {
                 let fun = create_physical_fun(fun)?;
                 (fun)(&inputs)
             }
-            ScalarFunctionDefinition::UDF(ref fun) => fun.invoke(&inputs),
+            ScalarFunctionDefinition::UDF(ref fun) => {
+                let output = fun.invoke(&inputs)?;
+                let output_count = match &output {
```

Comment:
https://github.com/apache/datafusion/blob/19356b26f515149f96f9b6296975a77ac7260149/datafusion/functions/src/utils.rs#L108

I think we should check it inside the function wrapper

Reply:
then anyone who implement ScalarUDFImpl trait and not using this util can miss this validation right?

Reply:
for example this function: https://github.com/apache/datafusion/blob/b2cbc7809ee0656099169307a73aadff23ab1030/datafusion-examples/examples/advanced_udf.rs#L96

Reply:
> then anyone who implement ScalarUDFImpl trait and not using this util can miss this validation right?

True, but in another way, all the function is forced to check the length, even if it does not expect to. 🤔

Reply:
i think the initial purpose of this validation is to announce users who define their own udf to follow this constraint (maybe they provide a wrong implementation, and our code just panic with ambiguous error, for example the error in the reported issue https://github.com/apache/datafusion/issues/5635

after adding this validation, i found one udf violating this rule which is arrow_typeof.
Can we consider this function a udaf, i need your opinion here @alamb

Reply:
@duongcongtoai 
The reason why `arrow_typeof` is failing is that it does not consider the input is array or scalar but always return `ColumnarValue::Scalar`.

Reply:
The reason `arrow_typeof` failed is because it returns `ColumnValue::Scalar(ScalarValue::UTF8("type name"))`, so if the input is list with number of rows 7, we get 1 (because of ColumnValue::Scalar) but expected 7 in this case.

Change `arrow_typeof` to UDAF is definitely not a good idea.
In this case, I think we can add `validate_length` method to `ScalarUDFImpl`, disable `arrow_typeof` to skip the check for this special case

Reply:
It also address my previous concern that all functions are not forced to check the length. By default it does, if any function does not expect to like `arrow_typeof`, it is allowed to skip the check

Reply:
i wonder if i misunderstood the usage of the enum ColumnarValue::Scalar here. For example a input having 10 rows, and UDF return ColumnarValue::Scalar(somevalue), then this will be transformed into an output array with 10rows having the same value right?
Then i think this validation can only be applied if the output returned is of type ColumnarValue::Array wdyt?

Reply:
ahah confirmed: https://github.com/apache/datafusion/blob/e642cc2a94f38518d765d25c8113523aedc29198/datafusion/expr/src/columnar_value.rs#L57

Reply:
> then this will be transformed into an output array with 10rows having the same value right

But `into_array` is not called, so the result of `arrow_typeof` remains to be ColumnarValue::Scalar not Array.

Reply:
```
 but into_array is not called after invoke(), so the result of arrow_typeof remains to be a single row value ColumnarValue::Scalar
```
which is fine to skip the validation right? because such transformation (scalar -> expected output) will be done some where else, depends on the plan, for example during [projection](https://github.com/apache/datafusion/blob/e642cc2a94f38518d765d25c8113523aedc29198/datafusion/physical-plan/src/projection.rs#L315)

Reply:
> ColumnarValue::Scalar is used for single row

hmm, maybe this is the point i got confused, because the way i look at how this ColumnarValue work, this is not correct. e.g in projection stream above). How i understand is: ColumnarValue::Scalar looks like a single value represent (arbitrarily) a single shared value of n rows

Reply:
Skip the validation for `arrow_typeof ` is fine to me, but skip it for `ScalarValue` in general is not. I think most of the function expects scalar in scalar out, so we still need to check if the result is single row.

Reply:
thank you for the review :D understood

Reply:
Btw I also tested with Spark and Duckdb for their version of "typeof"
Duckdb
```
toai@ToaiPC:~/proj/rust/arrow-datafusion/datafusion/core/tests/data$ duckdb 
v0.10.2 1601d94f94
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
D select typeof(ints) from parquet_map.parquet
  ;
┌──────────────────────┐
│     typeof(ints)     │
│       varchar        │
├──────────────────────┤
│ MAP(VARCHAR, BIGINT) │
│ MAP(VARCHAR, BIGINT) │
│ MAP(VARCHAR, BIGINT) │
│ MAP(VARCHAR, BIGINT) │
│ MAP(VARCHAR, BIGINT) │
│ MAP(VARCHAR, BIGINT) │
```
and Spark
```
val data = Seq(
  ("Alice", Map("age" -> "25", "email" -> "alice@example.com")),
  ("Bob", Map("age" -> "30", "email" -> "bob@example.com")),
  ("Carol", Map("age" -> "35", "email" -> "carol@example.com"))
)
val df = data.toDF("name", "attributes")
val types = df.select($"name", typeof($"name").as("name_type"), $"attributes", typeof($"attributes").as("attributes_type"))
types.show(false)
+-----+---------+---------------------------------------+------------------+
|name |name_type|attributes                             |attributes_type   |
+-----+---------+---------------------------------------+------------------+
|Alice|string   |{age -> 25, email -> alice@example.com}|map<string,string>|
|Bob  |string   |{age -> 30, email -> bob@example.com}  |map<string,string>|
|Carol|string   |{age -> 35, email -> carol@example.com}|map<string,string>|
+-----+---------+---------------------------------------+------------------+
```
Look like they also respect number of input row

Reply:
oh actually current arrow_typeof also respect this behavior (added extra test case in arrow_typeof.slt)
```
query T
select arrow_typeof(col) from (select 1 as col union select 2 as col) as table_a;
----
Int64
Int64
```
=> So in this case ArrowTypeOfFunc always return a Scalar
```
        Ok(ColumnarValue::Scalar(ScalarValue::from(format!(
            "{input_data_type}"
        ))))
```
but the rendered record batch shows n rows anyway, does it mean wrong implementation with correct result?